File: chapter_3/identification/identify_plant.py
#!/usr/bin/env python3
"""
identify_plant.py  --  Chapter 3: Plant identification for qwen2.5:3b

Standalone -- no dependency on MATLAB, setup_plant.m, or any workspace.
Run directly:   python3 -u identify_plant.py

Identifies the three parameters of:
    l_mean = alpha * B  +  gamma * B^2  +  beta * q

Stage 1: B sweep at q=0  ->  fits alpha, gamma
Stage 2: sustained load sweep  ->  fits beta

Outputs:
    identified_params.json  (loadable in MATLAB via jsondecode)
    id_stage1.png
    id_stage2.png
"""

import sys
import time
import json
import threading
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import requests
from pathlib import Path
from datetime import datetime
from numpy.random import default_rng

# flush all prints immediately
print = lambda *a, **kw: __builtins__["print"](*a, **{**kw, "flush": True}) \
    if isinstance(__builtins__, dict) else \
    __import__("builtins").print(*a, **{**kw, "flush": True})

# -------------------------------------------------------------------------
# Configuration
# -------------------------------------------------------------------------
CFG = dict(
    url          = 'http://localhost:11434/api/generate',
    model        = 'qwen2.5:3b',
    # Use non-streaming with num_predict=1.
    # With 1 token output, end-to-end ≈ TTFT (generation term is ~50ms,
    # negligible vs prefill).  This matches the working MATLAB approach
    # and avoids streaming-buffer issues in Python.
    num_predict  = 1,
    timeout      = 20,      # s per request
    n_warmup     = 4,
    n_reps       = 5,       # repetitions per B in stage 1
    b_sweep      = [1, 2, 3, 4, 5, 6, 8, 10, 12],
    lambda_sweep = [2, 3, 5, 7, 8],
    settle_ticks = 25,
    measure_ticks= 15,
    b_max        = 12,
    b_min        = 1,
    q_max        = 20,
    out_dir      = Path(__file__).parent,
)

rng = default_rng(seed=42)


# -------------------------------------------------------------------------
# Single request latency (non-streaming, num_predict=1)
# -------------------------------------------------------------------------
def measure_latency(prompt: str) -> float:
    """
    Measure end-to-end latency for one request with num_predict=1.
    With 1 output token, end-to-end ≈ TTFT (prefill dominates).
    Returns wall-clock ms.  Returns penalty on any error.
    """
    body = {
        "model":   CFG["model"],
        "prompt":  prompt,
        "stream":  False,
        "options": {"num_predict": CFG["num_predict"]},
    }
    t0 = time.perf_counter()
    try:
        resp = requests.post(CFG["url"], json=body, timeout=CFG["timeout"])
        resp.raise_for_status()
        return (time.perf_counter() - t0) * 1000
    except Exception as e:
        return CFG["timeout"] * 1000   # penalty


def fire_concurrent(b: int, prompt: str = "What is 2+2?") -> list:
    """Fire b concurrent requests and return list of latencies [ms]."""
    results = [None] * b

    def worker(idx):
        results[idx] = measure_latency(prompt)

    threads = [threading.Thread(target=worker, args=(i,)) for i in range(b)]
    for t in threads: t.start()
    for t in threads: t.join()
    return results


# -------------------------------------------------------------------------
# Warm-up
# -------------------------------------------------------------------------
def warmup():
    print(f"Warming up {CFG['model']} ({CFG['n_warmup']} requests)...")
    lats = fire_concurrent(CFG["n_warmup"])
    valid = [l for l in lats if l < CFG["timeout"] * 900]
    print(f"  Latencies: {[f'{l:.0f}' for l in lats]} ms")
    if valid:
        print(f"  Mean warm latency: {np.mean(valid):.0f} ms")
    else:
        print("  WARNING: all warm-up requests timed out — is Ollama running?")
        sys.exit(1)
    print("  Warm-up complete.\n")


# =========================================================================
# STAGE 1
# =========================================================================
def stage1():
    print("=" * 60)
    print("STAGE 1: B sweep at q=0  (identifying alpha, gamma)")
    print("=" * 60 + "\n")

    b_vals  = CFG["b_sweep"]
    l_means = []
    l_stds  = []

    for b in b_vals:
        rep_means = []
        for rep in range(CFG["n_reps"]):
            lats  = fire_concurrent(b)
            valid = [l for l in lats if l < CFG["timeout"] * 900]
            rep_means.append(np.mean(valid) if valid else np.nan)
        mean_l = float(np.nanmean(rep_means))
        std_l  = float(np.nanstd(rep_means))
        l_means.append(mean_l)
        l_stds.append(std_l)
        print(f"  B={b:2d}:  mean = {mean_l:6.1f} ms  (std = {std_l:5.1f} ms)")

    l_means = np.array(l_means)
    b_arr   = np.array(b_vals, dtype=float)

    # l = alpha*B + gamma*B^2  ->  linear least squares
    A       = np.column_stack([b_arr, b_arr**2])
    params, _, _, _ = np.linalg.lstsq(A, l_means, rcond=None)
    alpha_id, gamma_id = float(params[0]), float(params[1])

    l_fit = A @ params
    ss_res = np.sum((l_means - l_fit)**2)
    ss_tot = np.sum((l_means - np.mean(l_means))**2)
    r2 = float(1 - ss_res / ss_tot) if ss_tot > 0 else 0.0

    print(f"\n  alpha = {alpha_id:.4f}  ms/req")
    print(f"  gamma = {gamma_id:.4f}  ms/req^2")
    print(f"  R^2   = {r2:.4f}\n")

    # Plot
    b_fine = np.linspace(1, max(b_vals), 200)
    fig, ax = plt.subplots(figsize=(7, 4))
    ax.errorbar(b_vals, l_means, yerr=l_stds, fmt='bo',
                capsize=4, label='Measured (mean ± std)', zorder=5)
    ax.plot(b_fine, alpha_id*b_fine + gamma_id*b_fine**2, 'r-',
            linewidth=2,
            label=f'α·B + γ·B²  (α={alpha_id:.3f}, γ={gamma_id:.4f})')
    ax.set_xlabel('Batch size B  [requests]')
    ax.set_ylabel('Latency  [ms]')
    ax.set_title(f'Stage 1 — B sweep at q=0  (qwen2.5:3b)\nR²={r2:.4f}')
    ax.legend(); ax.grid(True); fig.tight_layout()
    out = CFG["out_dir"] / "id_stage1.png"
    fig.savefig(out, dpi=150); plt.close(fig)
    print(f"  Plot saved: {out}\n")

    return alpha_id, gamma_id, r2, {
        "b_sweep": b_vals,
        "l_means": l_means.tolist(),
        "l_stds":  l_stds,
    }


# =========================================================================
# STAGE 2
# =========================================================================
def stage2(alpha_id: float, gamma_id: float):
    print("=" * 60)
    print("STAGE 2: Sustained load sweep  (identifying beta)")
    print("=" * 60 + "\n")

    q_ss_list, b_ss_list, l_ss_list = [], [], []

    for lam in CFG["lambda_sweep"]:
        print(f"  lambda = {lam} req/tick ...")
        q_k     = float(lam)
        lat_buf = [250.0] * 20
        buf_idx = 0
        q_hist, b_hist, l_hist = [], [], []
        total   = CFG["settle_ticks"] + CFG["measure_ticks"]

        for tick in range(total):
            a_k = int(rng.poisson(lam))
            b_k = int(np.clip(round(q_k + a_k), CFG["b_min"], CFG["b_max"]))

            t_wall = time.perf_counter()
            lats   = fire_concurrent(b_k)
            wall   = (time.perf_counter() - t_wall) * 1000

            valid = [l for l in lats if l < CFG["timeout"] * 900]
            for l in valid:
                lat_buf[buf_idx % 20] = l
                buf_idx += 1
            l_meas = float(np.mean(lat_buf))

            q_k = float(np.clip(q_k + a_k - b_k, 0, CFG["q_max"]))
            q_hist.append(q_k); b_hist.append(b_k); l_hist.append(l_meas)

            if tick % 5 == 0:
                print(f"    tick {tick:3d}/{total}  q={q_k:.1f}  "
                      f"b={b_k}  l={l_meas:.0f}ms  wall={wall:.0f}ms")

        meas = slice(CFG["settle_ticks"], total)
        q_ss = float(np.mean(q_hist[meas]))
        b_ss = float(np.mean(b_hist[meas]))
        l_ss = float(np.mean(l_hist[meas]))
        q_ss_list.append(q_ss); b_ss_list.append(b_ss); l_ss_list.append(l_ss)
        print(f"    => q_ss={q_ss:.2f}  b_ss={b_ss:.2f}  l_ss={l_ss:.1f} ms\n")

    q_ss = np.array(q_ss_list)
    b_ss = np.array(b_ss_list)
    l_ss = np.array(l_ss_list)

    service  = alpha_id * b_ss + gamma_id * b_ss**2
    residual = l_ss - service
    valid    = q_ss > 0.1

    if valid.sum() > 0:
        beta_id = float(
            (q_ss[valid] @ residual[valid]) / (q_ss[valid] @ q_ss[valid])
        )
    else:
        beta_id = 2.0
        print("  WARNING: no valid q_ss > 0.1 — falling back to beta=2.0")

    l_fit2  = service + beta_id * q_ss
    ss_res2 = np.sum((l_ss[valid] - l_fit2[valid])**2)
    ss_tot2 = np.sum((l_ss[valid] - np.mean(l_ss[valid]))**2)
    r2      = float(1 - ss_res2 / ss_tot2) if ss_tot2 > 0 else 0.0

    print(f"  beta  = {beta_id:.4f}  ms/req")
    print(f"  R^2   = {r2:.4f}\n")

    # Plot
    fig, axes = plt.subplots(1, 2, figsize=(12, 4))
    ax = axes[0]
    ax.scatter(q_ss, residual, c='blue', s=70, zorder=5)
    q_fine = np.linspace(0, max(q_ss)*1.1 if len(q_ss) else 5, 100)
    ax.plot(q_fine, beta_id*q_fine, 'r-', lw=2,
            label=f'β·q  (β={beta_id:.3f})')
    ax.set_xlabel('q_ss  [requests]')
    ax.set_ylabel('l_ss − (α·B + γ·B²)  [ms]')
    ax.set_title(f'Stage 2 — Queuing term\nβ={beta_id:.4f} ms/req,  R²={r2:.4f}')
    ax.legend(); ax.grid(True)

    ax = axes[1]
    sc = ax.scatter(b_ss, l_ss, c=CFG["lambda_sweep"][:len(b_ss)],
                    cmap='viridis', s=80, zorder=5)
    plt.colorbar(sc, ax=ax, label='lambda')
    b_fine2 = np.linspace(1, max(b_ss)*1.1 if len(b_ss) else 6, 100)
    for qi in [0, 1, 2, 3]:
        ax.plot(b_fine2, alpha_id*b_fine2 + gamma_id*b_fine2**2 + beta_id*qi,
                '--', lw=1, label=f'q={qi}')
    ax.set_xlabel('B_ss  [requests]')
    ax.set_ylabel('l_ss  [ms]')
    ax.set_title('Identified surface l(B, q)')
    ax.legend(fontsize=8); ax.grid(True)
    fig.tight_layout()
    out = CFG["out_dir"] / "id_stage2.png"
    fig.savefig(out, dpi=150); plt.close(fig)
    print(f"  Plot saved: {out}\n")

    return beta_id, r2, {
        "lambda_sweep": CFG["lambda_sweep"],
        "q_ss": q_ss.tolist(), "b_ss": b_ss.tolist(), "l_ss": l_ss.tolist(),
    }


# =========================================================================
# Main
# =========================================================================
def main():
    print("=" * 64)
    print(f"  Plant identification: {CFG['model']}")
    print("=" * 64 + "\n")

    warmup()
    alpha_id, gamma_id, r2_s1, raw_s1 = stage1()
    beta_id,  r2_s2,    raw_s2        = stage2(alpha_id, gamma_id)

    dt, tau_out = 1.0, 30.0
    z_cl         = np.exp(-dt / tau_out)
    K_il_id      = (z_cl - 1) / beta_id
    K_il_assumed = (z_cl - 1) / 2.0

    print("=" * 64)
    print(f"  IDENTIFIED PARAMETERS  ({CFG['model']})")
    print("=" * 64)
    print(f"  alpha = {alpha_id:8.4f}  ms/req       (R² stage 1 = {r2_s1:.4f})")
    print(f"  gamma = {gamma_id:8.4f}  ms/req^2     (R² stage 1 = {r2_s1:.4f})")
    print(f"  beta  = {beta_id:8.4f}  ms/req       (R² stage 2 = {r2_s2:.4f})")
    print(f"  ---")
    print(f"  vs assumed:  alpha=0.1  gamma=0.8  beta=2.0")
    print(f"  K_il identified = {K_il_id:.6f}")
    print(f"  K_il assumed    = {K_il_assumed:.6f}")
    print(f"  Ratio           = {K_il_id/K_il_assumed:.2f}x")
    print("=" * 64 + "\n")

    result = {
        "model":      CFG["model"],
        "timestamp":  datetime.now().isoformat(),
        "alpha":      alpha_id,
        "gamma":      gamma_id,
        "beta":       beta_id,
        "r2_stage1":  r2_s1,
        "r2_stage2":  r2_s2,
        "K_il_identified": K_il_id,
        "K_il_assumed":    K_il_assumed,
        "raw_stage1": {k: (v.tolist() if hasattr(v,'tolist') else v)
                       for k,v in raw_s1.items()},
        "raw_stage2": {k: (v.tolist() if hasattr(v,'tolist') else v)
                       for k,v in raw_s2.items()},
    }

    out_json = CFG["out_dir"] / "identified_params.json"
    with open(out_json, "w") as f:
        json.dump(result, f, indent=2)
    print(f"Saved: {out_json}")
    print(f"\nTo use in setup_plant.m:")
    print(f"  id = jsondecode(fileread('{out_json}'));")
    print(f"  perturbed.alpha = id.alpha;   % {alpha_id:.4f}")
    print(f"  perturbed.gamma = id.gamma;   % {gamma_id:.4f}")
    print(f"  perturbed.beta  = id.beta;    % {beta_id:.4f}")


if __name__ == "__main__":
    main()

File: chapter_3/identification/identify_plant.m
%% identify_plant.m  --  Chapter 3: Plant identification for qwen2.5:3b
%
% Standalone — no dependency on setup_plant.m or any workspace variable.
% Run this script independently to measure real TTFT from Ollama and fit
% the three model parameters:
%
%   l_mean = alpha*B + gamma*B^2 + beta*q
%
% Strategy (two sequential stages, no parallel pool):
%
%   Stage 1 -- B sweep at q=0
%     Single-shot bursts (no sustained arrivals, queue stays empty).
%     Fits alpha and gamma from l(B) = alpha*B + gamma*B^2.
%
%   Stage 2 -- Sustained load sweep
%     Run sustained arrivals at multiple lambda values, let queue settle,
%     then measure steady-state (B_ss, q_ss, l_ss).
%     Residual: l_ss - (alpha*B_ss + gamma*B_ss^2) = beta*q_ss
%     Fits beta by linear regression through the origin.
%
% Outputs
%   identified_params.mat   -- struct with alpha, gamma, beta, R2 values
%   id_stage1.png           -- TTFT vs B fit plot
%   id_stage2.png           -- queuing term vs q_ss fit plot

clear; clc;

% -------------------------------------------------------------------------
% Paths  (self-contained -- resolve relative to this file's location)
% -------------------------------------------------------------------------
id_dir  = fileparts(mfilename('fullpath'));
src_dir = fullfile(id_dir, '..', 'src');
addpath(src_dir);

fprintf('[init] id_dir  = %s\n', id_dir);
fprintf('[init] src_dir = %s\n', src_dir);

% Confirm ollama_ttft.m is reachable
if ~exist('ollama_ttft', 'file')
    error('ollama_ttft.m not found on path. Check src_dir = %s', src_dir);
end
fprintf('[init] ollama_ttft found on path\n\n');

% -------------------------------------------------------------------------
% Configuration
% -------------------------------------------------------------------------
cfg.url         = 'http://localhost:11434/api/generate';
cfg.model       = 'qwen2.5:3b';
cfg.num_predict = 1;      % 1 token: minimises generation component in TTFT
cfg.timeout     = 20;     % s per request (sequential, so generous)

% Stage 1
cfg.b_sweep     = [1 2 3 4 5 6 8 10 12];
cfg.n_reps      = 3;      % bursts per B value (averaged)

% Stage 2
cfg.lambda_vals  = [2 3 5 7 9];   % req/tick — each drives a different q_ss
cfg.settle_ticks = 20;            % ticks to reach steady state
cfg.measure_ticks= 15;            % ticks to average over
cfg.b_min        = 1;
cfg.b_max        = 12;
cfg.q_max        = 20;

fprintf('[config] model       = %s\n',   cfg.model);
fprintf('[config] b_sweep     = %s\n',   num2str(cfg.b_sweep));
fprintf('[config] n_reps      = %d\n',   cfg.n_reps);
fprintf('[config] lambda_vals = %s\n',   num2str(cfg.lambda_vals));
fprintf('[config] settle/meas = %d / %d ticks\n\n', cfg.settle_ticks, cfg.measure_ticks);

% -------------------------------------------------------------------------
% Confirm Ollama is reachable before starting
% -------------------------------------------------------------------------
fprintf('[check] Pinging Ollama at %s ...\n', cfg.url);
try
    opts = weboptions('Timeout', 5);
    webread('http://localhost:11434/api/tags', opts);
    fprintf('[check] Ollama is up.\n\n');
catch
    error('Ollama not reachable at localhost:11434. Is it running?');
end

% -------------------------------------------------------------------------
% Helper: single blocking TTFT call (no parfeval)
% -------------------------------------------------------------------------
    function lat_ms = single_ttft(url, model, prompt, n_pred, timeout_sec)
        % Direct Java streaming call — no parallel pool needed.
        timeout_ms = int32(timeout_sec * 1000);
        body = sprintf( ...
            '{"model":"%s","prompt":"%s","stream":true,"options":{"num_predict":%d}}', ...
            model, prompt, n_pred);
        url_obj = java.net.URL(url);
        conn    = url_obj.openConnection();
        conn.setRequestMethod('POST');
        conn.setRequestProperty('Content-Type', 'application/json');
        conn.setDoOutput(true);
        conn.setConnectTimeout(timeout_ms);
        conn.setReadTimeout(timeout_ms);
        out = conn.getOutputStream();
        out.write(int8(body));
        out.flush();
        out.close();
        t_start = tic;
        reader  = java.io.BufferedReader(java.io.InputStreamReader(conn.getInputStream()));
        reader.readLine();   % blocks until first token = TTFT
        lat_ms  = toc(t_start) * 1000;
        reader.close();
        conn.disconnect();
    end

% -------------------------------------------------------------------------
% Warm-up: 3 single requests to load model weights into GPU memory
% -------------------------------------------------------------------------
fprintf('[warmup] Firing 3 warm-up requests (sequential)...\n');
for w = 1:3
    t = tic;
    lat = single_ttft(cfg.url, cfg.model, 'Hello', 1, cfg.timeout);
    fprintf('  warmup %d: %.0f ms  (elapsed %.1f s)\n', w, lat, toc(t));
end
fprintf('[warmup] Done.\n\n');

% =========================================================================
% STAGE 1 -- B sweep at q = 0  (fit alpha, gamma)
% =========================================================================
fprintf('═══════════════════════════════════════════════════════\n');
fprintf('STAGE 1: B sweep at q=0  (fitting alpha and gamma)\n');
fprintf('═══════════════════════════════════════════════════════\n\n');

% Each "burst" fires B sequential requests immediately one after another.
% Because no arrivals come between requests the queue stays at 0.
% We average over n_reps bursts to reduce noise.

n_b      = numel(cfg.b_sweep);
l_mean_b = zeros(n_b, 1);
l_std_b  = zeros(n_b, 1);

for bi = 1:n_b
    b = cfg.b_sweep(bi);
    fprintf('[stage1] B = %2d  (%d reps x %d requests) ...\n', b, cfg.n_reps, b);

    rep_means = zeros(cfg.n_reps, 1);
    for r = 1:cfg.n_reps
        lats = zeros(1, b);
        for i = 1:b
            lats(i) = single_ttft(cfg.url, cfg.model, 'What is 2+2?', ...
                cfg.num_predict, cfg.timeout);
        end
        rep_means(r) = mean(lats);
        fprintf('  rep %d: mean TTFT = %.1f ms  [%s]\n', r, rep_means(r), ...
            num2str(round(lats), '%d '));
    end

    l_mean_b(bi) = mean(rep_means);
    l_std_b(bi)  = std(rep_means);
    fprintf('  --> B=%2d average: %.1f ms  (std=%.1f ms)\n\n', b, l_mean_b(bi), l_std_b(bi));
end

% Least-squares fit: l = alpha*B + gamma*B^2
fprintf('[stage1] Fitting l = alpha*B + gamma*B^2 ...\n');
A1       = [cfg.b_sweep(:), cfg.b_sweep(:).^2];
params1  = A1 \ l_mean_b;
alpha_id = params1(1);
gamma_id = params1(2);
l_fit_b  = A1 * params1;
r2_s1    = 1 - sum((l_mean_b - l_fit_b).^2) / sum((l_mean_b - mean(l_mean_b)).^2);

fprintf('[stage1] alpha = %.4f  ms/req\n',   alpha_id);
fprintf('[stage1] gamma = %.4f  ms/req^2\n', gamma_id);
fprintf('[stage1] R^2   = %.4f\n\n',         r2_s1);

% Plot
fig1 = figure('Name','Stage 1','Visible','off');
errorbar(cfg.b_sweep, l_mean_b, l_std_b, 'bo', 'MarkerFaceColor','b', 'LineWidth',1.2);
hold on;
b_fine = linspace(1, max(cfg.b_sweep), 200);
plot(b_fine, alpha_id*b_fine + gamma_id*b_fine.^2, 'r-', 'LineWidth', 2);
xlabel('B  [req]'); ylabel('TTFT  [ms]');
title(sprintf('Stage 1 — B sweep\nalpha=%.3f  gamma=%.4f  R^2=%.3f', ...
    alpha_id, gamma_id, r2_s1));
legend('Measured', 'Fit: \alpha B + \gamma B^2', 'Location','northwest');
grid on;
s1_path = fullfile(id_dir, 'id_stage1.png');
saveas(fig1, s1_path);
fprintf('[stage1] Plot saved: %s\n\n', s1_path);

% =========================================================================
% STAGE 2 -- Sustained load sweep  (fit beta)
% =========================================================================
fprintf('═══════════════════════════════════════════════════════\n');
fprintf('STAGE 2: Sustained load  (fitting beta)\n');
fprintf('═══════════════════════════════════════════════════════\n\n');

% Simple drain rule (NOT the cascade controller):
%   B[k] = clamp(round(q[k] + lambda), b_min, b_max)
% This drives the system to a non-trivial steady-state queue for each lambda.
%
% Rolling buffer for l_mean: 10-sample window.
% Queue state tracked as a software counter.

n_lam    = numel(cfg.lambda_vals);
q_ss_vec = zeros(n_lam, 1);
b_ss_vec = zeros(n_lam, 1);
l_ss_vec = zeros(n_lam, 1);
beta_vec = zeros(n_lam, 1);

for li = 1:n_lam
    lam = cfg.lambda_vals(li);
    fprintf('[stage2] lambda = %d req/tick\n', lam);
    fprintf('  settling for %d ticks, then measuring %d ticks ...\n', ...
        cfg.settle_ticks, cfg.measure_ticks);

    q_k     = lam;                       % start near natural equilibrium
    lat_buf = 200 * ones(1, 10);         % 10-sample rolling buffer, warm-start
    buf_idx = 0;

    total_ticks = cfg.settle_ticks + cfg.measure_ticks;
    q_log = zeros(total_ticks, 1);
    b_log = zeros(total_ticks, 1);
    l_log = zeros(total_ticks, 1);

    for tick = 1:total_ticks
        % Poisson arrivals
        a_k = poissrnd(lam);

        % Drain rule: absorb arrivals + drain a bit of queue
        b_k = min(cfg.b_max, max(cfg.b_min, round(q_k + a_k)));

        % Fire b_k SEQUENTIAL requests (no parallel pool)
        lats_tick = zeros(1, b_k);
        for i = 1:b_k
            lats_tick(i) = single_ttft(cfg.url, cfg.model, ...
                'What is 2+2?', cfg.num_predict, cfg.timeout);
        end

        % Update rolling buffer
        for i = 1:b_k
            buf_idx = mod(buf_idx, 10) + 1;
            lat_buf(buf_idx) = lats_tick(i);
        end
        l_meas = mean(lat_buf);

        % Queue update
        q_k = max(0, min(cfg.q_max, q_k + a_k - b_k));

        q_log(tick) = q_k;
        b_log(tick) = b_k;
        l_log(tick) = l_meas;

        % Print every tick so you can see progress
        phase = 'settle';
        if tick > cfg.settle_ticks; phase = 'MEAS  '; end
        fprintf('  tick %3d [%s] a=%d b=%d q=%.1f l=%.1f ms\n', ...
            tick, phase, a_k, b_k, q_k, l_meas);
    end

    % Steady-state averages over measurement window
    idx_meas = (cfg.settle_ticks+1):total_ticks;
    q_ss = mean(q_log(idx_meas));
    b_ss = mean(b_log(idx_meas));
    l_ss = mean(l_log(idx_meas));

    q_ss_vec(li) = q_ss;
    b_ss_vec(li) = b_ss;
    l_ss_vec(li) = l_ss;

    % Beta estimate from this operating point
    svc = alpha_id * b_ss + gamma_id * b_ss^2;
    if q_ss > 0.1
        beta_vec(li) = (l_ss - svc) / q_ss;
        fprintf('  --> q_ss=%.2f  b_ss=%.2f  l_ss=%.1f ms  svc=%.1f ms  beta_est=%.4f\n\n', ...
            q_ss, b_ss, l_ss, svc, beta_vec(li));
    else
        beta_vec(li) = NaN;
        fprintf('  --> q_ss=%.2f (too small, skipping beta estimate)\n\n', q_ss);
    end
end

% Fit beta: linear regression through origin on residual vs q_ss
fprintf('[stage2] Fitting beta by regression: (l_ss - svc) = beta * q_ss ...\n');
valid     = ~isnan(beta_vec) & q_ss_vec > 0.1;
svc_all   = alpha_id * b_ss_vec + gamma_id * b_ss_vec.^2;
residuals = l_ss_vec - svc_all;
beta_id   = (q_ss_vec(valid)' * residuals(valid)) / ...
            (q_ss_vec(valid)' * q_ss_vec(valid));

l_fit_s2  = svc_all + beta_id * q_ss_vec;
r2_s2     = 1 - sum((l_ss_vec(valid) - l_fit_s2(valid)).^2) / ...
                sum((l_ss_vec(valid) - mean(l_ss_vec(valid))).^2);

fprintf('[stage2] beta = %.4f  ms/req\n', beta_id);
fprintf('[stage2] R^2  = %.4f\n\n',       r2_s2);

% Plot
fig2 = figure('Name','Stage 2','Visible','off');
scatter(q_ss_vec, residuals, 60, 'b', 'filled');
hold on;
q_fine = linspace(0, max(q_ss_vec)*1.2, 100);
plot(q_fine, beta_id * q_fine, 'r-', 'LineWidth', 2);
xlabel('q_{ss}  [req]');
ylabel('l_{ss} - (\alpha B + \gamma B^2)  [ms]');
title(sprintf('Stage 2 — Queuing term\nbeta=%.4f ms/req   R^2=%.3f', beta_id, r2_s2));
legend('Measured', sprintf('\\beta q  (\\beta=%.4f)', beta_id));
grid on;
s2_path = fullfile(id_dir, 'id_stage2.png');
saveas(fig2, s2_path);
fprintf('[stage2] Plot saved: %s\n\n', s2_path);

% =========================================================================
% SUMMARY
% =========================================================================
fprintf('╔══════════════════════════════════════════════════════════╗\n');
fprintf('║  IDENTIFIED PARAMETERS  (%s)\n', cfg.model);
fprintf('╠══════════════════════════════════════════════════════════╣\n');
fprintf('║  alpha = %8.4f  ms/req        R^2 stage1 = %.4f\n', alpha_id, r2_s1);
fprintf('║  gamma = %8.4f  ms/req^2      R^2 stage1 = %.4f\n', gamma_id, r2_s1);
fprintf('║  beta  = %8.4f  ms/req        R^2 stage2 = %.4f\n', beta_id,  r2_s2);
fprintf('╠══════════════════════════════════════════════════════════╣\n');
fprintf('║  Assumed (stochastic model):                            ║\n');
fprintf('║    alpha=0.1000  gamma=0.8000  beta=2.0000              ║\n');
fprintf('╠══════════════════════════════════════════════════════════╣\n');

dt_s     = 1.0;
tau_out  = 30.0;
z_cl     = exp(-dt_s / tau_out);
K_il_id  = (z_cl - 1) / beta_id;
K_il_old = (z_cl - 1) / 2.0;
fprintf('║  Outer loop gain (tau_out=%.0fs):\n', tau_out);
fprintf('║    K_il (identified) = %.6f\n', K_il_id);
fprintf('║    K_il (assumed)    = %.6f\n', K_il_old);
fprintf('║    Ratio             = %.2fx\n', K_il_id / K_il_old);
fprintf('╚══════════════════════════════════════════════════════════╝\n\n');

% Copy these lines into setup_plant.m:
fprintf('Paste into setup_plant.m:\n');
fprintf('  perturbed.alpha = %.4f;\n', alpha_id);
fprintf('  perturbed.gamma = %.4f;\n', gamma_id);
fprintf('  perturbed.beta  = %.4f;\n', beta_id);

% =========================================================================
% SAVE
% =========================================================================
identified.alpha     = alpha_id;
identified.gamma     = gamma_id;
identified.beta      = beta_id;
identified.r2_stage1 = r2_s1;
identified.r2_stage2 = r2_s2;
identified.model     = cfg.model;
identified.timestamp = char(datetime('now'));
identified.raw.b_sweep      = cfg.b_sweep;
identified.raw.l_mean_b     = l_mean_b;
identified.raw.lambda_vals  = cfg.lambda_vals;
identified.raw.q_ss         = q_ss_vec;
identified.raw.b_ss         = b_ss_vec;
identified.raw.l_ss         = l_ss_vec;

out_path = fullfile(id_dir, 'identified_params.mat');
save(out_path, 'identified');
fprintf('\n[save] Results saved to: %s\n', out_path);
